modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/management_bar/components/CustomViewsControls.js

```diff
@@ -0,0 +1,126 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import ClayButton from '@clayui/button';
+import ClayDropDown from '@clayui/drop-down';
+import ClayForm, {ClayInput} from '@clayui/form';
+import ClayIcon from '@clayui/icon';
+import {ManagementToolbar} from 'frontend-js-components-web';
+import {openModal} from 'frontend-js-web';
+import React, {useContext, useState} from 'react';
+
+import FrontendDataSetContext from './../../FrontendDataSetContext';
+
+const CustomViewsControls = () => {
+	const [viewsDropdownActive, setViewsDropdownActive] = useState(false);
+	const [actionsDropdownActive, setActionsDropdownActive] = useState(false);
+
+	const {namespace} = useContext(FrontendDataSetContext);
+
+	const SaveCustomViewModalBody = () => {
+		return (
+			<ClayForm.Group>
+				<label htmlFor={`${namespace}customViewNameInput`}>
+					{Liferay.Language.get('name')}
+
+					<RequiredMark />
+				</label>
+
+				<ClayInput
+					autoFocus={true}
+					id={`${namespace}customViewNameInput`}
+					type="text"
+				/>
+			</ClayForm.Group>
+		);
+	};
+
+	const openSaveCustomViewModal = () => {
+		openModal({
+			bodyComponent: SaveCustomViewModalBody,
+			buttons: [
+				{
+					displayType: 'secondary',
+					label: Liferay.Language.get('cancel'),
+					type: 'cancel',
+				},
+				{
+					label: Liferay.Language.get('save'),
+					onClick: ({processClose}) => {
+						processClose();
+					},
+				},
+			],
+			title: Liferay.Language.get('save-new-view'),
+		});
+	};
```

Comment:
Can we define those components/ functions outside the main component closure too? It can make things more clear and easy to read IMHO

Reply:
It's always difficult to define what is more readable. It's subjective. For me, there are a few things I consider:
- Is this a helper function / variable / component, or it has a standalone purpose? Is is used multiple times? If it has a standalone purpose, it should be in file root, outside of all other contexts.
- Are we running into source formatting ugliness, caused by line length limitation? If the code is deep in hierarchy, it might make sense to move it out just to 'buy space'. If not, placing it inline code might be the best option.
- How interconnected is the component to the context? Would we need to push in multiple props? This adds potentially needless code bloat and complexity down the line. I've seen a lot of code where people added prop defaults or changed prop names in child components, making code harder to read. See for example, [FDS actions](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/actions/Actions.js#L177-L218). Just try to follow props on all the `onClicks`,  `handleClick`s and `handleAction`s.
- How complex is the code? The more complex it is, the more it makes sense to fully move out. The simpler it it, the less impact it makes to readability of outside context.

In general, I am more open to move components outside, as they usually better fit other criteria above. Helper functions, on the other side, tend to be simpler, more fitting to be internal. If they are not, it may be best to move them to a separate file, to be an exported utility.

Ideally for me, the root of the file would only contains exported code. But, sometimes, for the sake of readability, I do admit it is better to have helper code moved out. See, for example, how I recently extracted [the helper `Card` component](https://github.com/liferay/liferay-portal/blob/master/modules/apps/frontend-data-set/frontend-data-set-web/src/main/resources/META-INF/resources/views/cards/Cards.js#L56). The only purpose of that was to buy space, removing six indentations. I moved it fully outside because it was a complex component, it would disrupt readability of `Cards` if it was inside it.

In this PR:
- Buying space is the only purpose of `openSaveCustomViewModal`. It is a helper function. It will not be used multiple times or externally. It is highly dependent on the context, makes no sense outside it. It is reasonably simple. If not for buying space, this code would normally belong inline. With that in mind, I would strongly prefer this to be inside the main component.
- As for `SaveCustomViewModalBody`, it's main purpose is also to buy space. It is also highly dependent on the context. Although currently there is only one prop, `namespace`, there will likely be a lot more in the future. It should stay reasonably simple in HTML, but the complexity in generating custom view objects will grow. Still, we may only pass in callback to it, keeping props small. I don't have a strong opinion on this one, it both makes sense external or internal.

Reply:
Thanks for explaining Marko, <3